sycl/doc/extensions/proposed/sycl_ext_oneapi_address_cast.asciidoc

```diff
@@ -101,7 +101,8 @@ constexpr inline access::address_space generic_space = access::address_space::ge
 
 template <access::address_space Space,
           typename ElementType>
-multi_ptr<ElementType, Space, access::decorated::no>
+multi_ptr<std::remove_pointer_t<remove_decoration_t<ElementType *>>, Space,
+          access::decorated::no>
 static_address_cast(ElementType* ptr);
```

Comment:
I'm not convinced this is a good idea.  If I understand correctly, this will allow you to cast a raw decorated pointer to a `multi_ptr` that is is a *different* address space.  Why would we want to allow this?  For example, I think the following code would compile with no diagnostic:

```
int __attribute__((opencl_local)) *p;            // Raw decorated pointer to local address space
auto mp = static_address_cast<global_space>(p);  // Cast to global address space
```

We already have a safe way to create a `multi_ptr` from a raw decorated pointer that is in the *correct* address space:

```
int __attribute__((opencl_local)) *p;   // Raw decorated pointer to local address space
multi_ptr<int, local_space> mp{p};      // Create a multi_ptr in local address space
```

This is enabled through this constructor:

```
explicit multi_ptr(
    typename multi_ptr<ElementType, Space, access::decorated::yes>::pointer);
```

Reply:
> Why would we want to allow this?

We won't, it will result in a `static_assert`. Primary motivation is to allow

```
int __attribute__((opencl_local)) *p;
auto mp = static_address_cast<generic_space>(p);
```

for the use in templated code.

My current PR has the same `static_assert` for the `dynamic_address_cast`, but I'm now thinking I  should just return `nullptr` from `if constexpr`, so that the uses like https://github.com/intel/llvm/blob/dae7032284d78d9f30676fc07348f9475eebb7e9/sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp#L217-L227 could be switched to it.

Reply:
It seems like it would be better to just allow a generic-space `multi_ptr` to be constructed from any sort of raw pointer:

```
// Available only when: Space == access::address_space::generic_space
explicit multi_ptr(
    typename multi_ptr<ElementType, access::address_space::global_space, access::decorated::yes>::pointer);
explicit multi_ptr(
    typename multi_ptr<ElementType, access::address_space::local_space, access::decorated::yes>::pointer);
explicit multi_ptr(
    typename multi_ptr<ElementType, access::address_space::private_space, access::decorated::yes>::pointer);
explicit multi_ptr(ElementType *);
```

Then, you could do this:

```
int __attribute__((opencl_local)) *p;
multi_ptr<int, generic_space> mp{p};
```

> We won't, it will result in a static_assert.

Yes, I see now that the implementation does this.  However, the specification does not say it is required.

Reply:
After offline discussion with @gmlueck , decided that the change in the specification isn't needed. Created https://github.com/intel/llvm/pull/15435 to remove `TODO`s in the implementation instead.

@Pennycook - FYI in case you disagree with the new direction.

Reply:
I think I agree, but I want to clarify something...

> My current PR has the same `static_assert` for the `dynamic_address_cast`, but I'm now thinking I should just return `nullptr` from `if constexpr`.

Are you still planning to do this?  I think we should, because this was always the intent.  Note that the specification encourages a diagnostic for `static_address_cast`, but doesn't mention one for `dynamic_address_cast`.

My hope was that we could start to rewrite all the functions that look/behave like `group_load` such that they all contain dispatch based on `dynamic_address_cast`.  We can then start to teach the compiler to optimize away the dispatch, either because certain cases return `nullptr` at compile-time (when we have address space information) or because a later pass can perform address space propagation and optimize away certain calls to `OpGenericCastToPtr`.

Reply:
> Are you still planning to do this?

I've only introduced it in this PR, so it's not in `intel/llvm` and there is nothing to remove.

Reply:
Sure, but you could still add an explicit branch to return `nullptr` after a `constexpr` check, right?  Even if this PR isn't merged, we'll still generate an `OpGenericCast` instruction even when we know at compile-time that we don't need one.

I thought you were suggesting something like:
```c++
if constexpr (SrcAS != Space && Space != generic_space) {
  return ret_ty(nullptr);
}
```